### PR TITLE
[Kernel-Spark][DSv2] Wireframe for UCManagedSnapshotManager

### PR DESCRIPTION
**Pull request by TimothyW553**
## 🥞 Stacked PR
Use this [link](https://github.com/delta-io/delta/pull/5655/files/07f3b5b1a0e9f9eb1794ac4b7b7ff151cc9a3b10..0b5886852bbe9871ab2c6ee5893da940ef69edcc) to review incremental changes.
- [catalogtableutils-ccv2](https://github.com/delta-io/delta/pull/5477) [[Files changed](https://github.com/delta-io/delta/pull/5477/files)]
  - [**stack/snapshotmanager-wireframe**](https://github.com/delta-io/delta/pull/5655) [[Files changed](https://github.com/delta-io/delta/pull/5655/files/07f3b5b1a0e9f9eb1794ac4b7b7ff151cc9a3b10..0b5886852bbe9871ab2c6ee5893da940ef69edcc)]
    - [stack/snapshotmanager-implementation](https://github.com/delta-io/delta/pull/5677) [[Files changed](https://github.com/delta-io/delta/pull/5677/files/0b5886852bbe9871ab2c6ee5893da940ef69edcc..a7b46bbc8d16040a10cc61fa65876af5b9e6ed0b)]
      - [stack/snapshotmanager-factory-wireup](https://github.com/delta-io/delta/pull/5678) [[Files changed](https://github.com/delta-io/delta/pull/5678/files/a7b46bbc8d16040a10cc61fa65876af5b9e6ed0b..6538c8b460050666e918f8853f67ab70ccea2d92)]

---------

#### Which Delta project/connector is this regarding?

- [ ] Spark
- [ ] Standalone
- [ ] Flink
- [ ] Kernel
- [ ] Other (fill in here)

## Description

This PR:
  - Adds `kernelUnityCatalog` as a dependency for `sparkV2` to wire in UC-managed functionality.
  - Introduces `UCManagedSnapshotManager` skeleton for UC-managed tables that will delegate snapshot/
    commit operations to UCCatalogManagedClient (methods currently TODO).

(Orange is the previous PR, green is what we've introduced in this PR. Routing is deferred)
```mermaid
flowchart TB
    A["Spark table CatalogTable"] -- decide snapshot manager --> B{"Select snapshot manager"}
    B -- path based --> P["PathBasedSnapshotManager loadLatestSnapshot loadSnapshotAt"]
    C["UCCatalogManagedClient"] -- build UC REST client --> D["UC REST client"]
    D -- GET ratified commits and table metadata --> U["Unity Catalog server"]
    X["UCUtils extractTableInfo<br>(tableId, path, uri, token)"] -- needs UC table details --> M["UCManagedSnapshotManager<br>(UCCatalogManagedClient(UCClient))"]
    B -- UC managed --> X
    M --> C

     C:::added
     X:::extraction
     M:::added
    classDef extraction fill:#ffb347,stroke:#d97706,color:#000
    classDef added fill:#b7e1a1,stroke:#2d6a4f,color:#000
```

## How was this patch tested?

Locally and CI.

## Does this PR introduce _any_ user-facing changes?
No.

### CODE REVIEW

File: kernel-spark/src/main/java/io/delta/kernel/spark/snapshot/unitycatalog/UCUtils.java [hunk 1]
```diff
@@ -29,8 +29,8 @@
 /**
  * Utility class for extracting Unity Catalog table information from Spark catalog metadata.
  *
- * <p>This class isolates Spark dependencies, allowing {@link UnityCatalogAdapter} to be created
```
Comment (huan233usc) [inline kernel-spark/src/main/java/io/delta/kernel/spark/snapshot/unitycatalog/UCUtils.java]: Can we make this change in the parent PR?

**Comment by huan233usc** [comment]
Plz update the description

File: kernel-spark/src/main/java/io/delta/kernel/spark/snapshot/unitycatalog/UCUtils.java (hunk 1 above)
Comment (timothyw553) [reply kernel-spark/src/main/java/io/delta/kernel/spark/snapshot/unitycatalog/UCUtils.java]: Done, updated.

File: kernel-spark/src/main/java/io/delta/kernel/spark/snapshot/unitycatalog/UCManagedSnapshotManager.java [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+import java.util.Optional;
+import org.apache.hadoop.conf.Configuration;
+
+/**
+ * Snapshot manager for Unity Catalog managed tables.
+ *
+ * <p>This implementation delegates to {@link UCCatalogManagedClient} for all snapshot and commit
+ * operations, leveraging UC's ratified commits as the source of truth.
+ */
+public class UCManagedSnapshotManager implements DeltaSnapshotManager {
```
Comment (raveeram-db) [inline kernel-spark/src/main/java/io/delta/kernel/spark/snapshot/unitycatalog/UCManagedSnapshotManager.java]: Would it be more clear to rename the class to `UCManagedTableSnapshotManager `

File: kernel-spark/src/main/java/io/delta/kernel/spark/snapshot/unitycatalog/UCManagedTableSnapshotManager.java [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+import io.delta.kernel.spark.exception.VersionNotFoundException;
+import io.delta.kernel.spark.snapshot.DeltaSnapshotManager;
+import io.delta.kernel.unitycatalog.UCCatalogManagedClient;
+import java.util.Optional;
+import org.apache.hadoop.conf.Configuration;
+
+/**
+ * Snapshot manager for Unity Catalog managed tables.
+ *
+ * <p>This implementation delegates to {@link UCCatalogManagedClient} for all snapshot and commit
```
Comment (tdas) [inline kernel-spark/src/main/java/io/delta/kernel/spark/snapshot/unitycatalog/UCManagedTableSnapshotManager.java]: i dont thnk we should mentioned internal implementation details. tomorrow if we change this, i doubt anyone will remember to update the docs.

rather than say 
- what is special about this SnapshotManager (catalogManaged feature, etc.)

File: kernel-spark/src/test/java/io/delta/kernel/spark/snapshot/unitycatalog/UCTableInfoTest.java
Comment (tdas) [inline kernel-spark/src/test/java/io/delta/kernel/spark/snapshot/unitycatalog/UCTableInfoTest.java:26]: I missed this in the previous PR... but we shoudl not add tests for each of these classes that are basically case classes with zero logic in them. these LLM generated stuff are bogus.

File: kernel-spark/src/main/java/io/delta/kernel/spark/snapshot/unitycatalog/UCManagedTableSnapshotManager.java (hunk 3 above)
Comment (timothyw553) [reply kernel-spark/src/main/java/io/delta/kernel/spark/snapshot/unitycatalog/UCManagedTableSnapshotManager.java]: Update javadoc - removed specific internal implementations and added significance of this UC Snapshot manager.

File: kernel-spark/src/main/java/io/delta/kernel/spark/snapshot/unitycatalog/UCManagedTableSnapshotManager.java [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+  public UCManagedTableSnapshotManager(
+      UCCatalogManagedClient ucCatalogManagedClient,
+      UCTableInfo tableInfo,
+      Configuration hadoopConf) {
+    this.ucCatalogManagedClient =
+        requireNonNull(ucCatalogManagedClient, "ucCatalogManagedClient is null");
+    requireNonNull(tableInfo, "tableInfo is null");
+    this.tableId = tableInfo.getTableId();
+    this.tablePath = tableInfo.getTablePath();
+    this.engine = DefaultEngine.create(requireNonNull(hadoopConf, "hadoopConf is null"));
```
Comment (scottsand-db) [inline kernel-spark/src/main/java/io/delta/kernel/spark/snapshot/unitycatalog/UCManagedTableSnapshotManager.java]: Ideally you don't save a reference to the engine - that can be expensive

Also -- what happens when the delta-spark dsv2 connector changes/creates its own new engine? Will you have a bunch of `DefaultEngine.create(...)` callsites that you have to replace?

Seems like it would be better to create a SparkKernelEngine.create(...) and then you can swap out the implementaiton in there in one centralized location, eh?

Comment (timothyw553) [reply kernel-spark/src/main/java/io/delta/kernel/spark/snapshot/unitycatalog/UCManagedTableSnapshotManager.java]: That makes sense - updated so the constructor takes `engine` parameter instead of `hadoopConf`, and the caller will be responsible for creating engine.

In a follow up PR, will centralize all these `DefaultEngine.create(...)` calls: https://github.com/delta-io/delta/issues/5675

File: build.sbt [hunk 5]
```diff
@@ -462,6 +462,7 @@ lazy val sparkV1Filtered = (project in file("spark-v1-filtered"))
 lazy val sparkV2 = (project in file("kernel-spark"))
   .dependsOn(sparkV1Filtered)
   .dependsOn(kernelDefaults)
+  .dependsOn(kernelUnityCatalog % "compile->compile;test->test")
```
Comment (tdas) [inline build.sbt:465]: ooooh.. the first integration between UC-Kernel + Spark! yay!
